lib/ansible/modules/cloud/vmware/vmware_guest_tools_upgrade.py

```diff
@@ -0,0 +1,184 @@
+            - 'Examples:'
+            - '   folder: /ha-datacenter/vm'
+            - '   folder: ha-datacenter/vm'
+            - '   folder: /datacenter1/vm'
+            - '   folder: datacenter1/vm'
+            - '   folder: /datacenter1/vm/folder1'
+            - '   folder: datacenter1/vm/folder1'
+            - '   folder: /folder1/datacenter1/vm'
+            - '   folder: folder1/datacenter1/vm'
+            - '   folder: /folder1/datacenter1/vm/folder2'
+            - '   folder: vm/folder2'
+            - '   folder: folder2'
+        default: /vm
+   datacenter:
+        description:
+            - Destination datacenter for the deploy operation
+        required: True
+extends_documentation_fragment: vmware.documentation
+author:
+    - Mike Klebolt (@MikeKlebolt) <michael.klebolt@centurylink.com>
+'''
+
+EXAMPLES = '''
+- name: Upgrade VMWare Tools
+  vmware_guest_facts:
+    hostname: 192.168.1.209
+    username: administrator@vsphere.local
+    password: vmware
+    datacenter: ha-datacenter
+    validate_certs: no
+    uuid: 421e4592-c069-924d-ce20-7e7533fab926
+  delegate_to: localhost
+'''
+
+RETURN = ''' # '''
+
+try:
+    import pyVmomi
+    from pyVmomi import vim
+except ImportError:
+    pass
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from ansible.module_utils.vmware import (PyVmomi, gather_vm_facts, vmware_argument_spec,
+                                         wait_for_task)
+
+
+class PyVmomiHelper(PyVmomi):
+    def __init__(self, module):
+        super(PyVmomiHelper, self).__init__(module)
+
+    def upgrade_tools(self, vm):
+        result = {'failed': False, 'changed': False, 'msg': ''}
+        # Exit if VMware tools is already up to date
+        if vm.guest.toolsStatus == "toolsOk":
+            result.update(
+                changed=False,
+                msg="VMware tools is already up to date",
+            )
+            return result
+
+        # Fail if VM is not powered on
+        elif vm.summary.runtime.powerState != "poweredOn":
+            result.update(
+                failed=True,
+                msg="VM must be powered on to upgrade tools",
+            )
+            return result
+
+        # Fail if VMware tools is either not running or not installed
+        elif vm.guest.toolsStatus in ["toolsNotRunning", "toolsNotInstalled"]:
+            result.update(
+                failed=True,
+                msg="VMware tools is either not running or not installed",
+            )
+            return result
+
+        # If vmware tools is out of date, check major OS family
+        # Upgrade tools on Linux and Windows guests
+        elif vm.guest.toolsStatus == "toolsOld":
+            try:
+                if vm.guest.guestFamily in ["linuxGuest", "windowsGuest"]:
+                    task = vm.UpgradeTools()
+                wait_for_task(task)
```

Comment:
wait_for_task returns status of task so it would be better if we catch it and pass it to return 

```python
changed, err_msg = wait_for_task(task)
result.update(changed=changed, msg=err_msg)
```